package/service/submission_service.go

```diff
@@ -13,17 +15,183 @@ type SubmissionService interface {
+		break
+	case "teacher":
+		submission_models, err = us.submissionRepository.GetAllForTeacher(tx, user.Id, queryParams)
+		break
+	}
+
+	if err != nil {
+		us.logger.Errorf("Error getting all submissions: %v", err.Error())
+		return nil, err
+	}
+
+	var result []schemas.Submission
+	for _, submission_model := range submission_models {
+		result = append(result, *us.modelToSchema(&submission_model))
+	}
+
+	return result, nil
+}
+
+func (us *SubmissionServiceImpl) GetById(tx *gorm.DB, submissionId int64, user schemas.UserSession) (schemas.Submission, error) {
+	submission_model, err := us.submissionRepository.GetSubmission(tx, submissionId)
+	if err != nil {
+		us.logger.Errorf("Error getting submission: %v", err.Error())
+		return schemas.Submission{}, err
+	}
+
+	switch user.Role {
+	case "admin":
+		// Admin is allowed to view all submissions
+		break
+	case "student":
+		// Student is only allowed to view their own submissions
+		if submission_model.UserId != user.Id {
+			us.logger.Errorf("User %v is not allowed to view submission %v", user.Id, submissionId)
+			return schemas.Submission{}, ErrPermissionDenied
+		}
+		break
+	case "teacher":
+		// Teacher is only allowed to view submissions for tasks they created
+		if submission_model.Task.CreatedBy != user.Id {
+			us.logger.Errorf("User %v is not allowed to view submission %v", user.Id, submissionId)
+			return schemas.Submission{}, ErrPermissionDenied
+		}
+		break
+	}
+
+	return *us.modelToSchema(submission_model), nil
+}
+
+func (us *SubmissionServiceImpl) GetAllForUser(tx *gorm.DB, userId int64, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error) {
+	submission_models, err := us.submissionRepository.GetAllByUserId(tx, userId, queryParams)
+	if err != nil {
+		us.logger.Errorf("Error getting all submissions for user: %v", err.Error())
+		return nil, err
+	}
+
+	switch user.Role {
+	case "admin":
+		// Admin is allowed to view all submissions
+		break
+	case "student":
+		// Student is only allowed to view their own submissions
+		if userId != user.Id {
+			us.logger.Errorf("User %v is not allowed to view submissions", user.Id)
+			return []schemas.Submission{}, ErrPermissionDenied
+		}
+		break
+	case "teacher":
+		// Teacher is only allowed to view submissions for tasks they created
+		for i, submission := range submission_models {
+			if submission.Task.CreatedBy != user.Id {
+				submission_models = append(submission_models[:i], submission_models[i+1:]...)
+			}
+		}
+		break
+	}
+
+	var result []schemas.Submission
+	for _, submission_model := range submission_models {
+		result = append(result, *us.modelToSchema(&submission_model))
+	}
+
+	return result, nil
+}
+
+func (us *SubmissionServiceImpl) GetAllForGroup(tx *gorm.DB, groupId int64, user schemas.UserSession, queryParams map[string][]string) ([]schemas.Submission, error) {
+	var err error
+	submission_models := []models.Submission{}
+
+	switch user.Role {
```

Comment:
https://go.dev/wiki/Switch#:~:text=Go's%20switch%20statements%20are%20pretty,the%20end%20of%20each%20case.